xorg.*: update, add update script

sdht0 opened the pull request:
## Description of changes

* Added an update script. Looks for updated versions and commits any changes. Reuses the existing perl script to regenerate nix expressions. Adapted from https://gist.github.com/NickCao/fa592237f768c32ff26688d99b20f81f by @NickCao.
* Updated packages to latest available versions using the script.
* Successfully built all packages. 
* xf86videosavage now builds so removed the broken flag.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lolbinarycat commented:
> Successfully built 4-5 of the packages.

do you mean 4/5? or are you unsure how many packages you successfully built?

sdht0 commented:
> do you mean 4/5? or are you unsure how many packages you successfully built?

Ah sorry for being imprecise. I tried building the updated packages at random, with some failing to build which I then fixed.

sdht0 commented:
@marsam gentle ping, since you merged the earlier xorg PRs.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3702

sdht0 commented:
Updated to add latest releases.

Of note: xorg-server has security updates: https://lists.x.org/archives/xorg-announce/2024-April/003497.html

If needed, xorg-server can be updated in a separate PR directly to master.

cc @vcunat

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3763

MostAwesomeDude reviewed (approved):
Obviously up to @vcunat as to merge timing, but this all looks decent. I appreciate the update script!

vcunat commented:
I'm not aware of any need to control merge timing.

marsam commented on pkgs/servers/x11/xorg/update.py:
```diff
@@ -0,0 +1,116 @@
+print("Updating tarballs.list...")
+
+with open("./tarballs.list", "w") as f:
+    for l in updated_tarballs:
+        f.write(f'{l}\n')
```
nitpick
```suggestion
    f.writelines(tarball + '\n' for tarball in updated_tarballs)
```

marsam commented on pkgs/servers/x11/xorg/update.py:
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env nix-shell
+#!nix-shell --pure --keep NIX_PATH -i python3 -p nix git python3Packages.packaging python3Packages.beautifulsoup4 python3Packages.requests
```
nitpick
```suggestion
#!nix-shell --pure --keep NIX_PATH -i python3 -p nix git "python3.withPackages (ps: [ ps. packaging ps.beautifulsoup4 ps.requests ])"
```

marsam commented on pkgs/servers/x11/xorg/update.py:
```diff
@@ -0,0 +1,116 @@
+changes_text = []
+for line in lines_tarballs:
+    line = line.rstrip("\n")
+    if any(line.startswith(frag) for frag in [mirror, lurl]):
```
I think we can use `continue` here:
```python
    if not line.startswith((mirror, lurl)):
        continue
    pname, rem = line.rsplit("-", 1)
    ...
```

sdht0 replied:
This was the bug. Lines should be always added to tarballs file, with or without changes.

marsam reviewed (commented):
Looks great! Thank you! 
I left a minor style suggestions. Besides that, LGTM!

sdht0 commented on pkgs/servers/x11/xorg/update.py:
```diff
@@ -0,0 +1,116 @@
+            not href.endswith(".tar.bz2")
+            and not href.endswith(".tar.gz")
+            and not href.endswith(".tar.xz")
+        ):
```
Ah nice thanks #til

sdht0 commented:
Thanks @marsam. I applied all your suggestions.

marsam commented on pkgs/servers/x11/xorg/default.nix:
```diff
@@ -3135,34 +3135,14 @@ self: with self; {
   })) {};
 
-  # THIS IS A GENERATED FILE.  DO NOT EDIT!
-  xf86videonouveau = callPackage ({ stdenv, pkg-config, fetchurl, xorgproto, libdrm, udev, libpciaccess, xorgserver, testers }: stdenv.mkDerivation (finalAttrs: {
```
Ofborg is failing to eval because `xf86videonouveau` was removed

sdht0 replied:
Yup that's what I get for not reviewing the diffs after modifying the script.

sdht0 committed:
xorg: add update script

sdht0 committed:
xorg.*: update

luit: 20230201 -> 20240102
editres: 1.0.8 -> 1.0.9
iceauth: 1.0.9 -> 1.0.10
listres: 1.0.5 -> 1.0.6
xdm: 1.1.14 -> 1.1.16
xev: 1.2.5 -> 1.2.6
xfontsel: 1.1.0 -> 1.1.1
xload: 1.1.4 -> 1.2.0
xlsfonts: 1.0.7 -> 1.0.8
xmessage: 1.0.6 -> 1.0.7
xpr: 1.1.0 -> 1.2.0
xrefresh: 1.0.7 -> 1.1.0
xsm: 1.0.5 -> 1.0.6
xorg-docs: 1.7.2 -> 1.7.3
xorg-sgml-doctools: 1.12 -> 1.12.1
xf86-video-chips: 1.4.0 -> 1.5.0
xf86-video-nv: 2.1.22 -> 2.1.23
xf86-video-savage: 2.4.0 -> 2.4.1
xf86-video-vboxvideo: 1.0.0 -> 1.0.1
encodings: 1.0.7 -> 1.1.0
libfontenc: 1.1.7 -> 1.1.8
libX11: 1.8.7 -> 1.8.9
libxcb: 1.16 -> 1.17.0
libXdmcp: 1.1.4 -> 1.1.5
libXmu: 1.1.4 -> 1.2.1
xcb-proto: 1.16.0 -> 1.17.0
xorgproto: 2023.2 -> 2024.1
gccmakedep: 1.0.3 -> 1.0.4
xorg-server: 21.1.12 -> 21.1.13

sdht0 committed:
xorg.xf86videosavage: mark as not broken anymore after update

marsam reviewed (approved):
LGTM. Thank you!
